refactor: pipes and wires

JustAPotota opened the pull request:
This refactors all of the block side of fluid pipes, wires, and their walkways - block entities and pipe/wire network code is for another time. Models have gotten an overhaul too, especially walkways.

Fixes #454, #411, #372

- [x] Separate fluid pipe blocks per color
  - Walkways will be separate PR
- [x] Consolidate pipe/wire shape code
- [x] Consolidate walkway code
- [x] Implement FluidLoggable on more things
  - [x] Fluid pipes
  - [x] Wires
  - [x] Walkways
- [x] New baked models
- [x] Item models
  - [x] Fluid pipes
  - [x] Wires
  - [x] Walkways
- [x] Proper translations
- [x] Stained pipe recipes

## TODO
- Fluid pipe colors don't prevent fluid flow

lingyun-create commented:
Why is the model of the fluid pipe divided into multiple files?

lingyun-create commented:
How about adding a feature，left-clicking on the connection part of two wires or pipes with a wrench will disconnect it?

JustAPotota commented:
> How about adding a feature，left-clicking on the connection part of two wires or pipes with a wrench will disconnect it?

I don't think it's necessary, the colored pipes are designed specifically for that use case of stopping pipes from connecting.

Roelymole reviewed (commented):
Blockbench exports models using tabs instead of spaces, would you be able to just replace them?

Two of the struts on the underside of the wire and fluid walkways have their uv values the wrong way round (note the positions of the darker pixels).

![2025-05-06_11 44 50](https://github.com/user-attachments/assets/0aa0e384-e614-48ac-bd75-e85024d591ac)

The item models need to be updated to match too.

![2025-05-06_11 50 06](https://github.com/user-attachments/assets/6ce35a9f-a20d-4948-90b2-571bb0eca16e)

Also, there seems to be some z-fighting with the wire walkways.

![2025-05-06_11 45 23](https://github.com/user-attachments/assets/0fea773a-bb14-4fca-b37a-438606de9fba)

JustAPotota commented:
> Blockbench exports models using tabs instead of spaces, would you be able to just replace them?

Didn't notice that, fixed now.

> Two of the struts on the underside of the wire and fluid pipe walkways have their uv values the wrong way round (note the positions of the darker pixels).
> 
> ![2025-05-06_11 44 50](https://private-user-images.githubusercontent.com/161453965/440741523-0aa0e384-e614-48ac-bd75-e85024d591ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NjQyNDcsIm5iZiI6MTc0NjU2Mzk0NywicGF0aCI6Ii8xNjE0NTM5NjUvNDQwNzQxNTIzLTBhYTBlMzg0LWU2MTQtNDhhYy1iZDc1LWU4NTAyNGQ1OTFhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQyMDM5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMWNlYTU3NGIwY2Y5MWM4MmI0NTIzZmVkMTQyZDhiZjk3YmUxMzY0MDExMjZmNWY4OWVmMDlmZjU2OGE1ZDNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nweBRyMLTdRcx2NtrUu8g2wk1Q5xy1v6IftAu_jldgM)

> Also, there seems to be some z-fighting with the wire walkways.
> 
> ![2025-05-06_11 45 23](https://private-user-images.githubusercontent.com/161453965/440741630-0fea773a-bb14-4fca-b37a-438606de9fba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NjQyNDcsIm5iZiI6MTc0NjU2Mzk0NywicGF0aCI6Ii8xNjE0NTM5NjUvNDQwNzQxNjMwLTBmZWE3NzNhLWJiMTQtNGZjYS1iMzdhLTQzODYwNmRlOWZiYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQyMDM5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTBmMzIxNGQwMDkyOGM1YzBlOGIwYmM3MmI4Y2QwYThmYTY1OTc1ZWY1ZGEwY2M4N2ZhZTBiZGQxMjVmYjQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.veGHtRHN2vcyOaklhpczNrlGgtxeqilHCkYEuhgficg)

Both fixed locally. I've basically remade the walkway models.

> The item models need to be updated to match too.
> 
> ![2025-05-06_11 50 06](https://private-user-images.githubusercontent.com/161453965/440742783-6ce35a9f-a20d-4948-90b2-571bb0eca16e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NjQyNDcsIm5iZiI6MTc0NjU2Mzk0NywicGF0aCI6Ii8xNjE0NTM5NjUvNDQwNzQyNzgzLTZjZTM1YTlmLWEyMGQtNDk0OC05MGIyLTU3MWJiMGVjYTE2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQyMDM5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZmE3NWE4ODk0N2QxZmU4ZWYwOGE4NjM5ZmEwN2I2MjJkOWE3MGYyM2U4NzZiMDBkM2YxYmM1OWVmYjQ4ODAyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.O4bg9hFFVgNtFL5-IYTVaUk_uyAKr8QjSdKfhimf1Xc)

Will take care of that later on. I'll redo the task list at some point since a lot has changed.

lingyun-create commented:
seems that their model is not very complete yet（expect for pipe's）
![Uploading 2025-05-17_10.49.11.png…]()
![Uploading 2025-05-17_10.49.36.png…]()
![2025-05-17_10 49 41](https://github.com/user-attachments/assets/72dd3780-d9ed-46d9-9411-be4b4a53af7c)
When there are blocks of the same type next to them, their model will be partially missing

JustAPotota commented:
> seems that their model is not very complete yet（expect for pipe's)
> When there are blocks of the same type next to them, their model will be partially missing

Only for wire walkways, since I haven't gotten around to refactoring wires yet. That's next on my list now that the walkways are done.

lingyun-create commented:
> > seems that their model is not very complete yet（expect for pipe's)
> > When there are blocks of the same type next to them, their model will be partially missing
> 
> Only for wire walkways, since I haven't gotten around to refactoring wires yet. That's next on my list now that the walkways are done.

Ordinary walkways have similar problems when they are next to wire walkways

JustAPotota commented:
My bad, turns out the cull faces on the walkway platform model were completely messed up. Let me know if you find any more issues with it.

lingyun-create commented:
Seems that wire ‘s item model needs  refactoring
![image](https://github.com/user-attachments/assets/14575ddd-d4e4-41fd-a74b-dc0ed1fcf003)
From certain perspectives,the model of pipewalkway is a bit strange and may need to be added with some extra detail to it
![image](https://github.com/user-attachments/assets/cd934d98-f053-4da8-a3cc-f791c410ca71)

JustAPotota commented:
Tentatively opening this up for review since there shouldn't be any more major changes. Still working on the last few things, see the top post for a list of known issues.

Roelymole commented:
The glass fluid pipes should probably listed in the creative inventory too.

Roelymole commented:
Another thing is that it might be nice to have recipes to craft the stained glass fluid pipes from stained glass panes.

![2025-05-19_23 22 27](https://github.com/user-attachments/assets/e43022a5-e0ed-4e9b-aa63-58d730da15b2)

Roelymole commented:
The sealable aluminum wires seem to have a ridiculously large voxel shape.

![2025-05-19_23 26 36](https://github.com/user-attachments/assets/2894046f-18ec-4ea9-a757-efbe0925857d)

Roelymole commented:
The connections are no longer being updated when changing the side configurations.

![2025-05-19_23 41 13](https://github.com/user-attachments/assets/905bc650-c96c-447b-b109-07f772f8e62e)

Roelymole commented:
The fluid pipes don't seem to transfer fuel. I don't know if this is an issue in general, or just for this pull request.

marcus8448 commented on src/main/java/dev/galacticraft/mod/api/block/PipeShapedBlock.java:
```diff
@@ -0,0 +1,180 @@
+            return;
+
+        if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+            level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
```
I might be reading the code incorrectly, but this feels really suspect to me. Wouldn't we want to specifically update the direction that got changed and not all the other faces? (the opposite of this method call)

JustAPotota replied:
Probably some leftover from me trying to get blocks to update correctly. I've removed both calls and everything still works.

marcus8448 commented on src/main/java/dev/galacticraft/mod/api/block/PipeShapedBlock.java:
```diff
@@ -0,0 +1,180 @@
+    protected @NotNull BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor levelAccessor, BlockPos pos, BlockPos neighborPos) {
+        if (levelAccessor instanceof Level level) {
+            if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+                level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
```
Same thing again, but with the added caveat that this additional update/check might be redundant if the `neighborChanged` check updates the changed direction.

Roelymole commented:
The wire walkways don't connect on the top face either.

![2025-05-20_00 07 05](https://github.com/user-attachments/assets/71add396-218b-47de-a468-198e0d98d3df)

marcus8448 commented on src/main/java/dev/galacticraft/mod/api/block/PipeShapedBlock.java:
```diff
@@ -0,0 +1,180 @@
+        return this.shapes[0];
+    }
+
+    public static VoxelShape[] makeShapes(float radius) {
```
I think this needs some special-casing for full-block pipes (e.g. sealable pipes/wires). Currently the bounds just explode out of the block.

JustAPotota replied:
My plan is to have the full-block ones as their own thing separate from the PipeShapedBlock set of classes, since the block entity is really the only thing they have in common with the thin pipe-likes.

marcus8448 commented on src/main/java/dev/galacticraft/mod/content/block/special/aluminumwire/tier2/HeavySealableAluminumWireBlock.java:
```diff
@@ -31,7 +31,7 @@
 public class HeavySealableAluminumWireBlock extends WireBlock {
     public HeavySealableAluminumWireBlock(Properties settings) {
-        super(settings);
+        super(1f, settings);
```
This is probably too large? (see `PipeShapedBlock#makeShapes` comment)

marcus8448 commented on src/main/java/dev/galacticraft/mod/content/block/special/aluminumwire/tier1/SealableAluminumWireBlock.java:
```diff
@@ -31,7 +31,7 @@
 public class SealableAluminumWireBlock extends WireBlock {
     public SealableAluminumWireBlock(Properties settings) {
-        super(settings);
+        super(0.5f, settings);
```
See `PipeShapedBlock#makeShapes` comment

marcus8448 commented on src/main/java/dev/galacticraft/mod/api/block/FluidPipeBlock.java:
```diff
@@ -0,0 +1,193 @@
+
+    @Nullable
+    @Override
+    public abstract PipeBlockEntity newBlockEntity(BlockPos pos, BlockState state);
```
Technically redundant as the type is specified in the class definition.

marcus8448 reviewed (changes requested):
Thanks for the PR! It looks really good - the inheritance structure/hierarchy is a lot nicer.

marcus8448 commented:
> The fluid pipes don't seem to transfer fuel. I don't know if this is an issue in general, or just for this pull request.

I think it's an issue in general - pipes/wires have lots of network/transfer issues

lingyun-create commented:
When will the (sealer) heavy wire be made

JustAPotota commented:
> When will the (sealable) heavy wire be made?

That'll be in another PR. I don't think we have textures for it yet?

JustAPotota commented:
I think I might call this PR here finished and let someone else figure out how to add color support to the pipe networking code.

marcus8448 reviewed (approved):
Looks good - thanks again!

Roelymole commented on src/main/java/dev/galacticraft/mod/data/tag/GCBlockTagProvider.java:
Could you add each stained glass fluid pipe to the corresponding dyed block tag?

For example, add the white stained glass fluid pipe to the `ConventionalBlockTags.WHITE_DYED` tag.

Roelymole reviewed (changes requested):
Apart from a few minor little inconsistencies this PR looks great.

Maybe recipes that require glass fluid pipes should accept any of the stained glass fluid pipes as well.

We could follow the conventions of the `#c:glass_blocks` and `#c:glass_blocks/colorless` block/item tags by having the `#galacticraft:glass_fluid_pipes` block/item tag with the subtags `#galacticraft:glass_fluid_pipes/colorless` and `#galacticraft:glass_fluid_pipes/stained` (or something similar).

JustAPotota commented:
> I think you have forgotten to track the src/main/generated/data/galacticraft/tags/item/glass_fluid_pipes.json and src/main/generated/data/galacticraft/tags/item/glass_fluid_pipes/stained.json files.

Oops, fixed now.

> I would prefer it if you also added block tags versions of the item tags, but if you don't feel like it I would be happy to do this myself later.

Yeah, that can be for another day. Maybe it can be incorporated into the pipe logic so that any pipes with the tag will connect with each other.
